esp/services/ws_access/ws_accessService.cpp

```diff
@@ -4009,6 +4025,467 @@ bool Cws_accessEx::onAccountPermissions(IEspContext &context, IEspAccountPermiss
+                    appendAccountPermission(newPermission, perm, accountPermissions, groupAccountPermissions);
+                }
+            }
+        }
+        bool doesPermissionAccountMatchThisAccount(CPermission &perm)
+        {
+            int accountType = perm.getAccount_type();
+            if (isGroupAccountReq && accountType == USER_ACT)
+                return false; //The account in the perm is not a group account.
+
+            const char *actName = perm.getAccount_name();
+            if (isEmptyString(actName))
+                return false;
+
+            //If the accountType matches with isGroupAccountReq, validate the actName.
+            if ((!isGroupAccountReq && (accountType == USER_ACT)) || (isGroupAccountReq && (accountType == GROUP_ACT)))
+                return streq(actName, accountNameReq); //The actName must match with the accountNameReq.
+
+            //Now, there is only one possibility left: isGroupAccountReq = false and accountType = GROUP_ACT.
+            //isGroupAccountReq = false: the AccountPermissionsForResource call is for an individual account.
+            //accountType = GROUP_ACT: the perm is for a group account; actName is the group name.
+            //We need to check whether the individual is a member of this group. 
+            return groupsAccountBelongsTo.find(actName) != NotFound;
+        }
+        IEspAccountPermission *createNewAccountPermission(const char *baseDNName,
+            const char *resourceName, CPermission &perm)
+        {
+            //Use the same code as in onAccountPermissions().
+            Owned<IEspAccountPermission> permission = createAccountPermission();
+            permission->setBasednName(baseDNName);
+            permission->setResourceName(resourceName);
+
+            int allows = perm.getAllows();
+            int denies = perm.getDenies();
+            if((allows & NewSecAccess_Access) == NewSecAccess_Access)
+                permission->setAllow_access(true);
+            if((allows & NewSecAccess_Read) == NewSecAccess_Read)
+                permission->setAllow_read(true);
+            if((allows & NewSecAccess_Write) == NewSecAccess_Write)
+                permission->setAllow_write(true);
+            if((allows & NewSecAccess_Full) == NewSecAccess_Full)
+                permission->setAllow_full(true);
+            if((denies & NewSecAccess_Access) == NewSecAccess_Access)
+                permission->setDeny_access(true);
+            if((denies & NewSecAccess_Read) == NewSecAccess_Read)
+                permission->setDeny_read(true);
+            if((denies & NewSecAccess_Write) == NewSecAccess_Write)
+                permission->setDeny_write(true);
+            if((denies & NewSecAccess_Full) == NewSecAccess_Full)
+                permission->setDeny_full(true);
+            return permission.getClear();
+        }
+        void appendAccountPermission(Owned<IEspAccountPermission> &permissionToBeAppended,
```

Comment:
Owned<IEspAccountPermission> &permissionToBeAppended,

This is an unusual type for a parameter.  It is generally only used if the parameter is going to be modified.  If it isn't then either

const Owned<IEspAccountPermission> &permissionToBeAppended,

or more commonly

IEspAccountPermission * permissionToBeAppended,

and LINK(permissionToBeAppended)